src/python/pants/util/retry.py

```diff
@@ -30,4 +30,5 @@ def retry_on_exception(func, max_retries, exception_types, backoff_func=lambda n
       return func()
     except exception_types as e:
       logger.debug('encountered exception on retry #{}: {!r}'.format(i, e))
-  raise
+      if i == max_retries - 1:
+        raise e
```

Comment:
That stray `raise` is meant to raise the last exception. It worked on Py2, but not Py3.

This fix captures the same logic in a way that works on both versions.